### PR TITLE
Ginkgo: Migrated 01-guestbook test

### PR DESCRIPTION
**Pull request by eloycoto**
- Migrated tests/k8s/tests/01-guestbook-test.sh to ginkgo
- Added the same policy in different namespace in a different test.

I didn't deprecated the 01-guestbook mainly because `k8sT/policies` are not validated yet. 

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

### CODE REVIEW

**Comment by eloycoto** [comment]
test-all-ginkgo

File: test/helpers/kubectl.go
Comment (ianvernon) [inline test/helpers/kubectl.go]: remove "it".
kubernetes --> Kubernetes.

File: test/helpers/kubectl.go
Comment (ianvernon) [inline test/helpers/kubectl.go]: remove "it".
kubernetes --> Kubernetes.

**Comment by ianvernon** [comment]
@eloycoto can you please tag as 'wip' and add 'pending-review' once the test has passed on Jenkins?

**Comment by eloycoto** [comment]
@ianvernon sorry, I should miss to add  something, it worked for me. sorry for bother!

**Comment by ianvernon** [comment]
@eloycoto no worries !

**Comment by eloycoto** [comment]
test-all-ginkgo

**Comment by eloycoto** [comment]
test-all-ginkgo

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
@ianvernon should be ok now ;-)

**Comment by ianvernon** [comment]
test-me-please

File: test/k8sT/Policies.go
Comment (ianvernon) [inline test/k8sT/Policies.go]: that check --> that checks

policyEnforcement --> policy enforcement

File: test/k8sT/Policies.go
Comment (ianvernon) [inline test/k8sT/Policies.go]: tested on --> tested in 

runtime/Policies.go

File: test/k8sT/Policies.go
Comment (ianvernon) [inline test/k8sT/Policies.go]: policy enforcement

File: test/k8sT/Policies.go
Comment (ianvernon) [inline test/k8sT/Policies.go]: Since calls to delete resources in Kubernetes are asynchronous, add checks to ensure that deletion occurred successfully, as we want to ensure that the test only ends when everything is cleaned up.

File: test/k8sT/Policies.go
Comment (ianvernon) [inline test/k8sT/Policies.go:516]: The bash test only imports the deprecated Redis policy if we are running Kubernetes 1.6. Won't we be importing the deprecated policy for each K8s version we use?

**Review by ianvernon — CHANGES_REQUESTED** [review]

File: test/k8sT/Policies.go
Comment (eloycoto) [reply test/k8sT/Policies.go:516]: I think that the both version should be tested.  Both should work in kubernetes 1.7 and we should be sure that it's tested correctly ;-)

**Comment by eloycoto** [comment]
test-all-ginkgo

File: test/k8sT/Policies.go
Comment (ianvernon) [inline test/k8sT/Policies.go]: using "err" is redundant; use `... : %q"), err)`

File: test/k8sT/Policies.go
Comment (ianvernon) [inline test/k8sT/Policies.go]: using "err" is redundant; use `... : %q"), err)`

File: test/k8sT/Policies.go
Comment (ianvernon) [reply test/k8sT/Policies.go:516]: You're right, I actually misread the bash test when looking side-by-side.

File: test/k8sT/Policies.go
Comment (ianvernon) [inline test/k8sT/Policies.go]: Please address my prior comment be sure to check for deletion of these resources propagated correctly to Cilium. Something like this is done in `Policies Across Namespaces` which is a test in this same file.

Comment (eloycoto) [reply test/k8sT/Policies.go]: Done!

**Comment by eloycoto** [comment]
test-all-ginkgo

**Comment by eloycoto** [comment]
test-all-ginkgo

**Comment by eloycoto** [comment]
test-all-ginkgo

**Comment by eloycoto** [comment]
test-all-ginkgo

**Comment by eloycoto** [comment]
test-all-ginkgo

**Comment by eloycoto** [comment]
test-all-ginkgo

**Comment by eloycoto** [comment]
test-all-ginkgo

**Comment by eloycoto** [comment]
test-all-ginkgo

**Comment by eloycoto** [comment]
test-all-ginkgo

File: test/helpers/cons.go
Comment (aanm) [inline test/helpers/cons.go:139]: Shouldn't we import this labels directly from cilium?

File: test/k8sT/Policies.go
Comment (aanm) [inline test/k8sT/Policies.go:98]: this is a TODO?

File: test/k8sT/Policies.go
Comment (aanm) [inline test/k8sT/Policies.go]: Can you make a little bit more descriptive?

File: test/k8sT/Policies.go
Comment (aanm) [inline test/k8sT/Policies.go]: Add a description `Should(BeNil(), fmt.Sprintf("Getting cilium pod on node %s", helpers.K8s1))`

File: test/k8sT/Policies.go
Comment (aanm) [inline test/k8sT/Policies.go]: Add description in case the success was not reached

**Comment by ianvernon** [comment]
@eloycoto please dismiss my review and re-assign me to review when this is ready. Thanks!

File: test/k8sT/Policies.go
Comment (ianvernon) [reply test/k8sT/Policies.go]: This is not done. Starting at line 405: https://github.com/cilium/cilium/pull/2648/files#diff-3e28e952665f25f47cc466d4d5c9a84cR405

The delete call will "succeed", but since Kubernetes is asynchronous, we need to wait for the resources that are being deleted to be cleaned up. This ensures that the test only exits once deletion of resources is completed. This provides a guarantee that the test environment has no leftovers from the test that just ran.

**Review by ianvernon — COMMENTED** [review]
Everything looks fine except for the checking that deletion of resources has indeed occurred - see my comment here https://github.com/cilium/cilium/pull/2648#discussion_r165182357

File: test/k8sT/Policies.go
Comment (aanm) [inline test/k8sT/Policies.go]: Why do we need the sleep?

File: test/k8sT/Policies.go
Comment (aanm) [inline test/k8sT/Policies.go]: Please more descriptive with the `It`

**Review by aanm — CHANGES_REQUESTED** [review]

File: test/helpers/cons.go
Comment (eloycoto) [reply test/helpers/cons.go:139]: It's a topic already discussed. If we use the cilium labels, on macOS can't using ginkgo. cc @ianvernon

File: test/k8sT/Policies.go
Comment (eloycoto) [reply test/k8sT/Policies.go:98]: nope

**Comment by eloycoto** [comment]
@ianvernon  I fixed the policy Delete, added a new helper function ;-)

**Comment by eloycoto** [comment]
test-all-ginkgo

File: test/helpers/cons.go
Comment (ianvernon) [reply test/helpers/cons.go:139]: For more context, see: https://github.com/cilium/cilium/issues/2308

File: test/k8sT/Policies.go
Comment (ianvernon) [inline test/k8sT/Policies.go]: This is halfway there, but still not quite there yet. I will clarify more what I think is needed here. Calls to `kubectl delete...` are asynchronous. The resources are not truly "deleted" after the call to `kubectl delete` returns. We need to wait until the pods are no longer present in the output of `kubectl get pods`, to ensure that the pods are fully cleaned up while this test is still running. I'm cautious about leftover cruft from prior tests interfering with test runs.

**Comment by eloycoto** [comment]
test-me-please

File: test/k8sT/Policies.go
Comment (ianvernon) [inline test/k8sT/Policies.go]: change to "Pods were not able to be deleted" - "finished" is unclear.

File: test/k8sT/Policies.go
Comment (ianvernon) [inline test/k8sT/Policies.go]: Typo - "termintate" --> "terminate"

**Review by ianvernon — APPROVED** [review]
Small nitpicks / typos. LGTM apart from those; since they are cosmetic in nature, you don't have to reassign to me for review.

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

File: test/helpers/cons.go
Comment (aanm) [reply test/helpers/cons.go:139]: #2702 addresses this, no?

**Review by aanm — APPROVED** [review]

File: test/helpers/cons.go
Comment (eloycoto) [reply test/helpers/cons.go:139]: yes!

**Comment by ianvernon** [comment]
@aanm created https://github.com/cilium/cilium/issues/2716 to handle the small new change to do after #2702 is merged. As janitor, am merging this since it has gotten both of our approvals. The Ginkgo tests failing in the last run of the build is due to a known issue that is not related to this branch.

**Pull request merged**
